staging/src/k8s.io/api/batch/v1/types.go

```diff
@@ -19,9 +19,14 @@ package v1
 import (
 	"k8s.io/api/core/v1"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
 )
 
-const JobCompletionIndexAnnotation = "batch.kubernetes.io/job-completion-index"
+const (
+	JobCompletionIndexAnnotation = "batch.kubernetes.io/job-completion-index"
+
+	JobTrackingFinalizer = "batch.kubernetes.io/job-tracking"
```

Comment:
rename since this is no longer being used as a finalizer, and add godoc explaining what object this is added to and why

Reply:
suggest framing this in terms of "X releases after GA of feature Y" instead of a specific release more than a year in the future

I'm also trying to envision the transition when this goes GA and we no longer need to decorate the job to get this behavior... is that sketched out anywhere, at what point the controller will unconditionally track pods this way? How will the controller decide it is safe to switch to using the finalizer pod mechanism on existing jobs that may already have pods created?

Reply:
Changed to "two releases after GA".

> I'm also trying to envision the transition when this goes GA and we no longer need to decorate the job to get this behavior.

I either have to find a safe way to add the finalizer to existing pods or we wait long enough to make sure that there shouldn't be existing Jobs that are tracked with the legacy behavior (or a combination of both).
Adding finalizers is tricky because we can't tell for sure if a finished pod doesn't have a finalizer because it's from the legacy tracking or because we counted it already. We can probably add the finalizer to running pods.

In our favor, we have that Jobs are run-to-completion. So eventually there shouldn't be non-finished Jobs with legacy behavior. Worst case scenario, we would recreate a few pods, once we get rid of the legacy tracking.

I was deferring these discussions to beta. I will add some notes when I update the KEP to match the implementation (the use of the finalizer as annotation and the extra work queue).

Reply:
> I was deferring these discussions to beta. I will add some notes when I update the KEP to match the implementation (the use of the finalizer as annotation and the extra work queue).

a clear picture of what will happen for this at each transition is important to have before we start adding annotations to things and implementing controllers

Reply:
> I either have to find a safe way to add the finalizer to existing pods or we wait long enough to make sure that there shouldn't be existing Jobs that are tracked with the legacy behavior (or a combination of both).

Honestly, that was what I was imagining that we'd keep both ways for more than usual supported time just to give folks the time to complete their jobs. All the new ones will always be tracked with the new mechanism when it gets beta, and old ones will eventually complete, as Aldo mentions. We just need to clearly define how long we'll keep the legacy behavior and explain that in the release notes.